feat: implement `get()` helper method.

sillyElvy opened the pull request:
## About

This PR tries to implement a get helper method.

## Checklist

- [x] I've ran `pre-commit` to format and lint the change(s) made.
- [x] I've checked to make sure the change(s) work on `3.8.6` and higher.
- [ ] This fixes/solves an [Issue](https://github.com/goverfl0w/discord-interactions/issues).
  - (If existent):
- [ ] I've made this pull request for/as: (check all that apply)
  - [ ] Documentation
  - [ ] Breaking change
  - [x] New feature/enhancement
  - [ ] Bugfix

sillyElvy committed:
feat: implement `bot.get()` method
resolve merge conflicts

sillyElvy committed:
feat: implement `bot.get()` method
resolve merge conflicts

sillyElvy committed:
ooops

sillyElvy commented:
Marking as draft since waiting for the cache...

Toricane commented on interactions/client/bot.py:
```diff
@@ -1134,6 +1139,59 @@ async def modify(
+        ...
+
+    async def get(self, obj: Type[_T], **kwargs) -> _T:
+        """
+        Gets an object from discord.
+        """
```
Probably change the docstring to this or something similar:
```py
r"""
A helper method for retrieving data from the Discord API in its object representation.

:param obj: The object to get.
:type obj: object
:param \**kwargs: The arguments to pass to the HTTP method.
:type \**kwargs: dict
:return: The object we're trying to get.
:rtype: object
"""
```

Toricane commented on interactions/client/bot.py:
```diff
@@ -1134,6 +1139,59 @@ async def modify(
+        Gets an object from discord.
+        """
+
+        if obj not in (Role, Emoji):
+            _func = getattr(self._http, f"get_{obj.__name__.lower()}")
+            _obj = await _func(**kwargs)
+            return obj(**_obj, _client=self._http)
+
+        else:
+            _guild = Guild(**await self._http.get_guild(kwargs.pop("guild_id")), _client=self._http)
+            _func = getattr(_guild, f"get_{obj.__name__.lower()}")
+            return await _func(**kwargs)
```
Inverse condition:
```py
if obj in (Role, Emoji):
    _guild = Guild(**await self._http.get_guild(kwargs.pop("guild_id")), _client=self._http)
    _func = getattr(_guild, f"get_{obj.__name__.lower()}")
    return await _func(**kwargs)
else:
    _func = getattr(self._http, f"get_{obj.__name__.lower()}")
    _obj = await _func(**kwargs)
    return obj(**_obj, _client=self._http)
```

Toricane reviewed (changes requested):
I'm late but here are some changes

FayeDel committed:
chore: Version bump.

sillyElvy committed:
resolve merge conflicts

sillyElvy committed:
resolve merge conflicts

sillyElvy committed:
resolve merge conflicts

sillyElvy committed:
resolve merge conflicts

pre-commit-ci[bot] committed:
ci: correct from checks.

sillyElvy committed:
revert: remove unnessecary line

sillyElvy committed:
ci

sillyElvy committed:
revert

Toricane commented on interactions/client/bot.py:
```diff
@@ -1416,6 +1421,33 @@ async def modify(
         return User(**data)
 
+    async def get(self, obj: Type[_T], **kwargs) -> _T:
+        """
+        A helper method for retrieving data from the Discord API in its object representation.
+
+        :param obj: The object to get. Should be a class object (not an instance!). For example: `interactions.Channel`.
+        :type obj: object
+        :param **kwargs: The arguments to pass to the HTTP method.
+        :type **kwargs: dict
+        :return: The object we're trying to get.
+        :rtype: object
+        """
```
```suggestion
        r"""
        A helper method for retrieving data from the Discord API in its object representation.

        :param obj: The object to get. Should be a class object (not an instance!). For example: `interactions.Channel`.
        :type obj: object
        :param \**kwargs: The arguments to pass to the HTTP method.
        :type \**kwargs: dict
        :return: The object we're trying to get.
        :rtype: object
        """
```

sillyElvy replied:
Flake 8 complained with the `\` before tho

Toricane replied:
That's how you do it for the docstring if you have **kwargs, it's also elsewhere, you can check the others out

sillyElvy replied:
I'll merge it and see if Flake8 complains again

Toricane commented on interactions/client/bot.py:
```diff
@@ -1416,6 +1421,33 @@ async def modify(
+        :rtype: object
+        """
+
+        if isinstance(obj, (Channel, Emoji, Guild, Member, Message, Role, Sticker, User)):
```
```suggestion
        if not isinstance(obj, type):
```

Toricane replied:
This checks if something is a class and not an instance [here is link](https://stackoverflow.com/questions/18499614/check-if-class-object)

Catalyst4222 replied:
That would do the opposite effect, where passing a class would raise the error instead of when an instance is passed. I do think that not hardcoding objects would be beneficial, though `inspect.isclass` would probably be the best way to do it

Toricane replied:
Oops forgot to add `not`

Toricane commented on interactions/client/bot.py:
```diff
@@ -1416,6 +1421,33 @@ async def modify(
+            _guild = Guild(**await self._http.get_guild(kwargs.pop("guild_id")), _client=self._http)
+            _func = getattr(_guild, _name)
+            return await _func(**kwargs)
+
+        else:
+            _func = getattr(self._http, _name)
+            _obj = await _func(**kwargs)
+            return obj(**_obj, _client=self._http)
```
You could technically remove the else condition if you want:
```suggestion
        _func = getattr(self._http, _name)
        _obj = await _func(**kwargs)
        return obj(**_obj, _client=self._http)
```

Toricane reviewed (changes requested):
Docstring and minor code improvements

i0bs reviewed (commented):
I've brought this up before both internally and in the support server, but doing `@overload` will eventually create a messy interface file. I think for type hinting purposes, the best clause is documenting `_T` to be a type of any data model that is recognised as a "resource" from the API.

sillyElvy commented:
> I've brought this up before both internally and in the support server, but doing `@overload` will eventually create a messy interface file. I think for type hinting purposes, the best clause is documenting `_T` to be a type of any data model that is recognised as a "resource" from the API.

But how should we tell the dev what attributes he needs?

AstreaTSS commented:
I don't think it's fully realistic to *completely* get rid of overloads, but I have an idea:

- One overload can have a bounded `TypeVar` that is bounded to the classes that need 2 attributes, and that class can typehint the two attributes needed, possibly `guild_id=..., id=...`.
- Another can be a normal `TypeVar` that only has one attribute, like `id=...`.

sillyElvy committed:
Update interactions/client/bot.py

Co-authored-by: Toricane <73972068+Toricane@users.noreply.github.com>

sillyElvy committed:
Update interactions/client/bot.py

Co-authored-by: Toricane <73972068+Toricane@users.noreply.github.com>

sillyElvy committed:
Update interactions/client/bot.py

Co-authored-by: Toricane <73972068+Toricane@users.noreply.github.com>

sillyElvy committed:
feat: add `@func.autocomplete` (#784)

* feat: add `@func.autocomplete`

* Update interactions/client/bot.pyi

Co-authored-by: Toricane <73972068+Toricane@users.noreply.github.com>

Co-authored-by: Toricane <73972068+Toricane@users.noreply.github.com>

LordOfPolls committed:
docs: rename mentions of dis-snek to NAFF (#791)

i0bs committed:
chore/ADMIN: force version onto old pypi

This is an executive decision made between Delta and I. We will be forcing a version 4.0 push onto our old branch to help transition old users using v3 to our new project. This does not necessarily mean this branch will now be receiving updates along with the rest. The goal is to make our current branch used as much as possible.

i0bs committed:
chore/ADMIN: force version onto old pypi

This is an executive decision made between Delta and I. We will be forcing a version 4.0 push onto our old branch to help transition old users using v3 to our new project. This does not necessarily mean this branch will now be receiving updates along with the rest. The goal is to make our current branch used as much as possible.

i0bs committed:
chore/ADMIN: fix content type

i0bs committed:
chore/ADMIN: fix content type

i0bs committed:
fix/ADMIN: correct improper description

i0bs committed:
fix/ADMIN: correct improper description

i0bs committed:
revert/ADMIN: change back to original

i0bs committed:
revert/ADMIN: change back to original

B1ue-Dev committed:
feat: Adding ``url`` property for getting emoji's url from the emoji's icon (#783)

* feat: Adding ``icon_url`` property for getting the guild's icon URL

* feat: Adding ``url`` property for ``Emoji`` for getting the emoji's URL

* ci: correct from checks.

* file: Update interactions/api/models/message.py

Co-authored-by: Max <maxyolo01.ytb@gmail.com>

* file: Update interactions/api/models/message.pyi

Co-authored-by: Max <maxyolo01.ytb@gmail.com>

* file: Update interactions/api/models/message.py

Co-authored-by: Max <maxyolo01.ytb@gmail.com>

* file: Update ./api/models/message.py to meet the review

* file: Update interactions/api/models/message.py

Co-authored-by: Toricane <73972068+Toricane@users.noreply.github.com>

Co-authored-by: pre-commit-ci[bot] <66853113+pre-commit-ci[bot]@users.noreply.github.com>
Co-authored-by: Max <maxyolo01.ytb@gmail.com>
Co-authored-by: Toricane <73972068+Toricane@users.noreply.github.com>

Toricane committed:
feat: ability to suppress embeds on message edit (#774)

* feat: ability to suppress embeds on message edit

* Update interactions/api/models/message.py

Co-authored-by: EdVraz <88881326+EdVraz@users.noreply.github.com>

* Update interactions/api/models/message.py

Co-authored-by: EdVraz <88881326+EdVraz@users.noreply.github.com>

Co-authored-by: EdVraz <88881326+EdVraz@users.noreply.github.com>

FayeDel committed:
chore/ADMIN: Reinclude image descriptions for GH reasons.

FayeDel committed:
chore/ADMIN: Reinclude image descriptions for GH reasons.

sillyElvy committed:
fix!: fix en locales (#796)

sillyElvy committed:
refactor: Add ``remove_commands`` argument for un- and reloading of extensions (#797)

* refactor: Add ``remove_commands`` argument for un- and reloading of extensions

* ci: correct from checks.

Co-authored-by: pre-commit-ci[bot] <66853113+pre-commit-ci[bot]@users.noreply.github.com>

Nanrech committed:
feat: Add clone_channel method to guild (#794)

* Add the things

* ci: correct from checks.

* Fix return in docstrings

* Remove redundancy

* ci: correct from checks.

* Update interactions/api/models/guild.py

Co-authored-by: Toricane <73972068+Toricane@users.noreply.github.com>

* Update interactions/api/models/guild.py

Co-authored-by: pre-commit-ci[bot] <66853113+pre-commit-ci[bot]@users.noreply.github.com>
Co-authored-by: Toricane <73972068+Toricane@users.noreply.github.com>

FayeDel committed:
feat: Implement Scheduled Events via Gateway. (#799)

* feat: Implement Scheduled Events via Gateway.

* docs: Document attributes to GuildScheduledEventUser object.

sillyElvy committed:
Merge remote-tracking branch 'origin/get()' into get()

# Conflicts:
#	interactions/client/bot.py
#	interactions/client/bot.pyi

sillyElvy commented:
I messed up the branch (again), I'll make a new PR
